apps/prairielearn/src/sprocs/ip_to_mode.sql

```diff
@@ -22,63 +20,66 @@ BEGIN
         mode := 'Public';
     END IF;
 
-    -- Does the user have an active PT reservation?
-    SELECT r.session_id
-    INTO v_session_id
-    FROM
-        pt_reservations AS r
-        JOIN pt_enrollments AS e ON (e.id = r.enrollment_id)
-    WHERE
-        e.user_id = ip_to_mode.authn_user_id
-        AND (date BETWEEN r.access_start AND r.access_end);
-
-    IF FOUND THEN
-        -- The user has a checked-in pt_reservation. Check whether the
-        -- reservation is in a testing center location that requires network
-        -- filtering.
-
-        SELECT loc.id, loc.filter_networks
-        INTO v_location_id, v_filter_networks
+    -- Consider each PT reservation which is either active or corresponds to
+    -- a session that will start soon or started recently.
+    FOR reservation IN
+        SELECT
+            r.session_id,
+            (ip_to_mode.date BETWEEN r.access_start AND access_end) AS reservation_active,
+            l.id AS location_id,
+            l.filter_networks AS location_filter_networks
         FROM
-            pt_sessions AS s
-            JOIN pt_locations AS loc ON (loc.id = s.location_id)
+            pt_reservations AS r
+            JOIN pt_enrollments AS e ON (e.id = r.enrollment_id)
+            JOIN pt_sessions AS s ON (s.id = r.session_id)
+            LEFT JOIN pt_locations AS l ON (l.id = s.location_id)
         WHERE
-            s.id = v_session_id;
-
-        IF FOUND THEN
-            -- The reservation is in a testing center location.
-            IF v_filter_networks THEN
-                -- The reservation is in a testing center location that requires
-                -- network filtering.
-
-                PERFORM *
-                FROM pt_location_networks AS ln
-                WHERE
-                    ln.location_id = v_location_id
-                    AND ip <<= ln.network;
-
-                IF FOUND THEN
-                    -- The user is physically inside the testing center. Set
-                    -- mode to 'Exam'.
-                    mode := 'Exam';
-                ELSE
-                    -- Although we have a checked-in reservation, the user is
-                    -- physically outside the testing center. Set mode to
-                    -- 'Public'.
-                    mode := 'Public';
-                END IF;
-            ELSE
-                -- The reservation is in a testing center location that doesn't
-                -- require network filtering. Set mode to 'Exam'.
+            e.user_id = ip_to_mode.authn_user_id
+            AND (
+                (r.access_end IS NULL and s.date BETWEEN ip_to_mode.date - '1 hour'::interval and ip_to_mode.date + '1 hour'::interval)
+                OR (ip_to_mode.date BETWEEN r.access_start AND r.access_end)
+            )
```

Comment:
Another condition we could check would be to activate Exam mode if we are shortly after `reservations.access_end`. This would prevent the case where, at the moment, a student finishes an exam due to time out and then instantly reverts back to Public mode, potentially giving a window of opportunity to exfiltrate some information via a Public assessment.

Reply:
Is it fair to say we'd only want to do this for IP-filtered sessions? Otherwise we'd run into a similar problem to what you described here: https://github.com/PrairieLearn/PrairieLearn/pull/9590#discussion_r1548829272. That is, a student could leave the session and find themselves unable to access PrairieLearn for a while.

I think skipping this check for non-IP-filtered sessions would be fine, as students could _already_ easily exfiltrate information in BYOD settings.

Reply:
I implemented this in 9379f27 by changing the `WHERE` clause to select reservations that ended within the last 30 minutes. It just sort of happened that this made it such that this would only change the behavior for IP-restricted sessions, as that's the only branch of code where we aren't also checking `reservation_active`.

Reply:
Looks good to me. I'd also fine with a longer time period here, but 30 minutes is probably fine in practice.

Reply:
We'll stick with 30 minutes for now, this is easy to adjust if needed.